README.md
# podsaver
an app to download episodes of your favourite podcast to your local archive

![Build Status](https://github.com/nekr0z/podsaver/actions/workflows/build.yml/badge.svg) [![codecov](https://codecov.io/gh/nekr0z/podsaver/branch/master/graph/badge.svg)](https://codecov.io/gh/nekr0z/podsaver) [![Go Report Card](https://goreportcard.com/badge/evgenykuznetsov.org/go/podsaver)](https://goreportcard.com/report/evgenykuznetsov.org/go/podsaver)

##### Table of Contents
* [Why](#why)
* [Usage](#usage)
* [How it works](#how-it-works)
* [Credits](#credits)

## Why
Sometimes things go offline for good, so some of us want a local archive of a favourite podcast to re-listen to during long winter nights.

## Usage
```
podsaver [-r] [-p path] url
```
The app will process the podcast feed (RSS or Atom) at a given `url`, scan the existing downloaded episodes in `path` (current directory by default) and try to download new and/or missing episodes.

## How it works
`podsaver` scans the local directory (current working directory if `-p path` is not given) and tries to guess podcast episode numbers from filenames (i.e. `episode14.mp3` or `14-new-guitar.avi` will be detected as episode 14 of the podcast). It then fetches the podcast feed from the given `url` and tries to match episodes in the feed to the latest one in the local directory. If matched, `podsaver` will download any episodes listed in the feed that are missing locally and save them as `number.ext` file (`ext` being extension such as `mp3` or `avi`) into the local directory. If given `-r` option, `podsaver` will also rename the existing files in a consistent manner.

For example, you have the following files:
```
episode3.mp3
episode 5.mp3
6.mp3
10.mp3
11-new-songs.mp3
awesomepodcast.13.mp3
```
and the feed has episodes 6 to 15 listed in it. `podsaver` will download what it can, so the folder will have:
```
episode3.mp3
episode 5.mp3
6.mp3
07.mp3
08.mp3
09.mp3
10.mp3
11-new-songs.mp3
12.mp3
awesomepodcast.13.mp3
14.mp3
15.mp3
```
`podsaver` has no way to get episodes 1, 2 and 4, but what was in the feed has been downloaded. Notice how the app appended trailing zeros to episodes 7, 8 and 9, because the latest episodes are in double-digit numbers.


Had you used `-r` option, existing files would be renamed, too:
```
03.mp3
05.mp3
06.mp3
07.mp3
08.mp3
09.mp3
10.mp3
11.mp3
12.mp3
13.mp3
14.mp3
15.mp3
```

## Credits
This software includes the following software or parts thereof:
* [Afero](https://github.com/spf13/afero) Copyright © 2014 Steve Francia
* [gofeed](https://github.com/mmcdole/gofeed) Copyright © 2016 mmcdole
* [gorilla/feeds](https://github.com/gorilla/feeds) Copyright © 2013-2018 The Gorilla Feeds Authors
* [The Go Programming Language](https://golang.org) Copyright 2009 The Go Authors


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Pre-release]

## [0.1.3] - 2022-04-06
### Changed
- spoof user agent while downloading an episode

## [0.1.2] - 2022-04-06
### Changed
- updated dependencies
- spoof user agent

## [0.1.1] - 2021-11-17
### Changed
- updated dependencies

## [0.1.0] - 2019-10-21
*initial release*

[Pre-release]: https://github.com/nekr0z/podsaver/releases/tag/latest
[0.1.3]: https://github.com/nekr0z/podsaver/releases/tag/v0.1.3
[0.1.2]: https://github.com/nekr0z/podsaver/releases/tag/v0.1.2
[0.1.1]: https://github.com/nekr0z/podsaver/releases/tag/v0.1.1
[0.1.0]: https://github.com/nekr0z/podsaver/releases/tag/v0.1.0


feed_test.go
// Copyright (C) 2019 Evgeny Kuznetsov (evgeny@kuznetsov.md)
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along tihe this program. If not, see <https://www.gnu.org/licenses/>.

package main

import (
	"fmt"
	"github.com/gorilla/feeds"
	"github.com/spf13/afero"
	"math/rand"
	"net/http"
	"net/http/httptest"
	"strconv"
	"testing"
	"time"
)

func TestMatchFeed(t *testing.T) {
	pod, server := generatePodcast(t, 15, 10, 10, 2)
	if err := pod.scanDir(); err != nil {
		t.Fatal(err)
	}

	if err := pod.matchFeed(); err != nil {
		t.Fatal(err)
	}

	server.Close()

	for i := 2; i <= 15; i++ {
		if pod.ep[i] == nil {
			t.Fatalf("episode %d does not exist", i)
		}
		if pod.ep[i].filename == "" {
			t.Fatalf("episode %d has no local filename", i)
		}
	}
	for i := 12; i <= 15; i++ {
		if pod.ep[i].filename != strconv.Itoa(i)+".pcast" {
			t.Fatalf("filename for episode %d (%s) does not match the expected (%s)", i, pod.ep[i].filename, strconv.Itoa(i)+".pcast")
		}
	}
}

func TestMatchFeedIncomplete(t *testing.T) {
	pod, server := generatePodcast(t, 15, 10, 2, 11)
	if err := pod.scanDir(); err != nil {
		t.Fatal(err)
	}

	if err := pod.matchFeed(); err != nil {
		t.Fatal(err)
	}

	server.Close()

	for i := 6; i <= 15; i++ {
		if pod.ep[i] == nil {
			t.Fatalf("episode %d does not exist", i)
		}
		if pod.ep[i].filename == "" {
			t.Fatalf("episode %d has no local filename", i)
		}
	}
	for i := 13; i <= 15; i++ {
		if pod.ep[i].filename != strconv.Itoa(i)+".pcast" {
			t.Fatalf("filename for episode %d (%s) does not match the expected (%s)", i, pod.ep[i].filename, strconv.Itoa(i)+".pcast")
		}
	}

	for i := 6; i <= 10; i++ {
		if pod.ep[i].filename != strconv.Itoa(i)+".pcast" {
			t.Fatalf("filename for episode %d (%s) does not match the expected (%s)", i, pod.ep[i].filename, strconv.Itoa(i)+".pcast")
		}
	}
}

func TestMatchFeedErrors(t *testing.T) {
	pod := podcast{}
	if err := pod.matchFeed(); err != errNoFeed {
		t.Fatal("expected error for podcast with no feed")
	}

	pod, server := generatePodcast(t, 5, 2, 2, 1)
	if err := pod.matchFeed(); err != errNotScanned {
		t.Fatalf("expected error - not scanned local dir")
	}

	if err := pod.scanDir(); err != nil {
		t.Fatal(err)
	}

	if err := pod.matchFeed(); err != errNoMatch {
		t.Fatalf("expected error: can not possibly match; actual error: %s", err)
	}

	pod.url = "http://localhost/notthere"
	if err := pod.matchFeed(); err == nil {
		t.Fatal("expected error for unreachable feed, got nil")
	} else if _, ok := err.(feedParseError); !ok {
		t.Fatalf("expected error for unreachable feed, got: %s", err)
	}

	pod.local = nil
	if err := pod.matchFeed(); err == nil || err == errNoMatch {
		t.Fatalf("expected error for lack of local fs")
	}
	server.Close()
}

func generatePodcast(t *testing.T, episodes, feedSize, downloaded, firstDownloaded int) (podcast, *httptest.Server) {
	t.Helper()

	fs := afero.NewMemMapFs()
	for i := 1; i <= episodes; i++ {
		filename := fmt.Sprintf("episode%02d.pcast", i)
		generateRandomFile(t, fs, filename, 2048)
	}

	lfs := afero.NewMemMapFs()
	for i := firstDownloaded; i < downloaded+firstDownloaded; i++ {
		filename := fmt.Sprintf("episode%02d.pcast", i)
		if err := copyFile(fs, lfs, filename, filename); err != nil {
			t.Fatal(err)
		}
	}

	rfs := afero.NewMemMapFs()
	if err := rfs.Mkdir("/http", 0755); err != nil {
		t.Fatal(err)
	}

	httpFs := afero.NewHttpFs(rfs)
	fileserver := http.FileServer(httpFs.Dir("/http"))
	server := httptest.NewServer(fileserver)
	now := time.Now()
	feed := &feeds.Feed{
		Title:       "mock podcast",
		Link:        &feeds.Link{Href: server.URL},
		Description: "test-test-test",
		Created:     now,
	}
	for i := 0; i < feedSize; i++ {
		filename := fmt.Sprintf("episode%02d.pcast", episodes-i)
		if err := copyFile(fs, rfs, filename, "/http/"+filename); err != nil {
			t.Fatal(err)
		}
		now = now.AddDate(0, 0, -14)
		feed.Add(&feeds.Item{
			Id:          strconv.Itoa(episodes - i),
			Link:        &feeds.Link{Href: server.URL},
			Title:       "yet another episode",
			Description: "latest and greatest",
			Enclosure: &feeds.Enclosure{
				Url:    fmt.Sprintf("%s/%s", server.URL, filename),
				Length: "1024",
				Type:   "media/podcast",
			},
			Created: now,
		})
	}
	rss, err := feed.ToRss()
	if err != nil {
		t.Fatal(err)
	}
	if err := afero.WriteFile(rfs, "/http/feed.rss", []byte(rss), 0644); err != nil {
		t.Fatal(err)
	}

	return podcast{
		local: lfs,
		url:   fmt.Sprintf("%s/feed.rss", server.URL),
	}, server
}

func generateRandomFile(t *testing.T, fs afero.Fs, name string, maxBytes int) {
	t.Helper()
	data := make([]byte, rand.Intn(maxBytes))
	if err := afero.WriteFile(fs, name, data, 0644); err != nil {
		t.Fatal(err)
	}
}

func TestCompareRemote(t *testing.T) {
	fs := afero.NewBasePathFs(afero.NewOsFs(), "testdata/")
	httpFs := afero.NewHttpFs(fs)
	fileserver := http.FileServer(httpFs.Dir("/downloaded"))
	server := httptest.NewServer(fileserver)
	local := afero.NewBasePathFs(fs, "/downloaded")

	ok, err := compareRemote(fmt.Sprintf("%s/749.mp3", server.URL), local, "15.mp3")
	if err != nil {
		t.Fatal(err)
	}
	if ok {
		t.Errorf("different files should differ")
	}

	ok, err = compareRemote(fmt.Sprintf("%s/15.mp3", server.URL), local, "15.mp3")
	if err != nil {
		t.Fatal(err)
	}
	if !ok {
		t.Errorf("same file should be the same")
	}

	server.Close()
}

func BenchmarkCompareRemote(b *testing.B) {
	fs := afero.NewBasePathFs(afero.NewOsFs(), "testdata/")
	httpFs := afero.NewHttpFs(fs)
	fileserver := http.FileServer(httpFs.Dir("/downloaded"))
	server := httptest.NewServer(fileserver)
	local := afero.NewBasePathFs(fs, "/downloaded")
	url := "https://audio.vgtrk.com/download?id=2441689"

	ok, err := compareRemote(url, local, "15.mp3")
	if err != nil {
		b.Fatal(err)
	}
	if ok {
		b.Errorf("different files should differ")
	}

	ok, err = compareRemote(url, local, "749.mp3")
	if err != nil {
		b.Fatal(err)
	}
	if !ok {
		b.Errorf("same file should be the same")
	}

	server.Close()
}

func TestFeedParseError(t *testing.T) {
	err := feedParseError{"none", errNotScanned}

	if s := err.Error(); s != "failed to read and parse feed from none" {
		t.Fatalf("wrong message: %s", s)
	}

	if e := err.Cause(); e != errNotScanned {
		t.Fatalf("want %s, got %s", errNotScanned, e)
	}
}


util_test.go
// Copyright (C) 2019 Evgeny Kuznetsov (evgeny@kuznetsov.md)
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along tihe this program. If not, see <https://www.gnu.org/licenses/>.

package main

import (
	"github.com/spf13/afero"
	"testing"
)

func TestCopyFileErrors(t *testing.T) {
	fs := afero.NewMemMapFs()

	if err := copyFile(fs, fs, "foo", "bar"); err == nil {
		t.Fatalf("should have returned error for non-existing file")
	}

	if err := afero.WriteFile(fs, "foo", nil, 0644); err != nil {
		t.Fatal(err)
	}

	fsr := afero.NewReadOnlyFs(fs)

	if err := copyFile(fs, fsr, "foo", "bar"); err == nil {
		t.Fatalf("should have returned error for writing on read-only file")
	}
}

func TestMostCurrent(t *testing.T) {
	testCases := []struct {
		epis []int
		max  int
	}{
		{epis: []int{1, 2, 3, 4, 5, 6}, max: 6},
		{epis: []int{6}, max: 6},
	}

	for _, testCase := range testCases {
		pod := podcast{ep: make(map[int]*episode)}

		for _, i := range testCase.epis {
			pod.ep[i] = &episode{}
		}

		max := pod.mostCurrent()

		if max != testCase.max {
			t.Errorf("want %d, got %d", testCase.max, max)
		}
	}
}


local_test.go
// Copyright (C) 2019 Evgeny Kuznetsov (evgeny@kuznetsov.md)
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along tihe this program. If not, see <https://www.gnu.org/licenses/>.

package main

import (
	"github.com/spf13/afero"
	"reflect"
	"sort"
	"testing"
)

type ep struct {
	n    int
	name string
}

func TestScanDir(t *testing.T) {
	var testCases = []struct {
		files []string
		epis  []ep
	}{
		{
			[]string{"1", "2", "3", "4", "5"},
			[]ep{
				{1, "1"},
				{2, "2"},
				{3, "3"},
				{4, "4"},
				{5, "5"},
			},
		},
		{
			[]string{"1.mp3", "2.mp3", "3.mp3", "5.mp3", "episode.7.mp3"},
			[]ep{
				{1, "1.mp3"},
				{2, "2.mp3"},
				{3, "3.mp3"},
				{5, "5.mp3"},
				{7, "episode.7.mp3"},
			},
		},
		{
			[]string{"01.mp3", "02.mp3", "25.mp3", "drop.txt", "26.mp3", "extras.mp3"},
			[]ep{
				{1, "01.mp3"},
				{2, "02.mp3"},
				{25, "25.mp3"},
				{26, "26.mp3"},
			},
		},
	}

	for _, testCase := range testCases {
		fs := populate(t, testCase.files)
		eps := scan(t, fs)
		assertEpisodes(t, eps, testCase.epis)
	}
}

func TestScanDirWithDirectories(t *testing.T) {
	files := []string{"100.avi", "101.avi", "102.mp3"}
	want := []ep{{100, "100.avi"}, {101, "101.avi"}, {102, "102.mp3"}}
	fs := populate(t, files)

	if err := fs.Mkdir("103", 0775); err != nil {
		t.Fatal(err)
	}

	eps := scan(t, fs)
	assertEpisodes(t, eps, want)
}

func TestScanDirError(t *testing.T) {
	pod := podcast{}

	if err := pod.scanDir(); err == nil {
		t.Error("scanned nothing")
	}

	pod.local = afero.NewBasePathFs(afero.NewMemMapFs(), "/bar/")

	if err := pod.scanDir(); err == nil {
		t.Error("scanned unreadable directory")
	}

	pod.local = afero.NewBasePathFs(afero.NewMemMapFs(), "/")

	if err := pod.scanDir(); err != errNoEpisodes {
		t.Error("expected error on empty directory")
	}
}

func TestRenameDownloaded(t *testing.T) {
	var testCases = []struct {
		files []string
		want  []string
	}{
		{
			files: []string{"6", "7", "8", "9", "10", "11"},
			want:  []string{"06", "07", "08", "09", "10", "11"},
		},
		{
			files: []string{"6", "70", "08", "196", "10", "11"},
			want:  []string{"006", "070", "008", "196", "010", "011"},
		},
		{
			files: []string{"65", "70", "0000867", "12345", "103", "1"},
			want:  []string{"00065", "00070", "00867", "12345", "00103", "00001"},
		},
		{
			files: []string{"6.html", "70.mp3", "08.avi", "196.pc", "10.txt", "11"},
			want:  []string{"006.html", "070.mp3", "008.avi", "196.pc", "010.txt", "011"},
		},
	}
	for _, testCase := range testCases {
		fs := populate(t, testCase.files)
		pod := podcast{local: fs}
		if err := pod.scanDir(); err != nil {
			t.Fatal(err)
		}

		if err := pod.renameDownloaded(); err != nil {
			t.Fatal(err)
		}

		var got []string
		for _, ep := range pod.ep {
			got = append(got, ep.filename)
		}
		assertStringSlices(t, got, testCase.want)

		eps := scan(t, fs)
		got = []string{}
		for _, ep := range eps {
			got = append(got, ep.name)
		}

		assertStringSlices(t, got, testCase.want)
	}
}

func populate(t *testing.T, filenames []string) afero.Fs {
	t.Helper()
	fs := afero.NewMemMapFs()

	for _, filename := range filenames {
		if err := afero.WriteFile(fs, filename, nil, 0644); err != nil {
			t.Fatal(err)
		}
	}

	return fs
}

func scan(t *testing.T, fs afero.Fs) []ep {
	t.Helper()
	pod := podcast{local: fs}

	if err := pod.scanDir(); err != nil {
		t.Fatal(err)
	}

	eps := make([]ep, len(pod.ep))

	i := 0
	for n, episode := range pod.ep {
		eps[i].n, eps[i].name = n, episode.filename
		i++
	}

	return eps
}

func assertEpisodes(t *testing.T, got, want []ep) {
	t.Helper()
	sort.Slice(got, func(i, j int) bool {
		return got[i].n < got[j].n
	})

	sort.Slice(want, func(i, j int) bool {
		return want[i].n < want[j].n
	})

	if !reflect.DeepEqual(got, want) {
		t.Errorf("want %v, got %v", want, got)
	}
}

func assertStringSlices(t *testing.T, got, want []string) {
	t.Helper()
	sort.Slice(got, func(i, j int) bool {
		return got[i] < got[j]
	})

	sort.Slice(want, func(i, j int) bool {
		return want[i] < want[j]
	})

	if !reflect.DeepEqual(got, want) {
		t.Errorf("want %v, got %v", want, got)
	}
}


feed.go
// Copyright (C) 2019 Evgeny Kuznetsov (evgeny@kuznetsov.md)
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along tihe this program. If not, see <https://www.gnu.org/licenses/>.

package main

import (
	"bytes"
	"fmt"
	"io"
	"mime"
	"net/http"
	"path/filepath"
	"sort"
	"strconv"
	"strings"

	"github.com/mmcdole/gofeed"
	"github.com/pkg/errors"
	"github.com/spf13/afero"
)

var (
	errNoFeed      = errors.New("no feed available")
	errNoMatch     = errors.New("could not match items in feed to local files")
	errNoEnclosure = errors.New("no enclosure or multiple enclosures for episode")
	errNotScanned  = errors.New("no data on downloaded episodes - not scanned?")
)

type feedParseError struct {
	url string
	err error
}

func (p feedParseError) Error() string {
	return fmt.Sprintf("failed to read and parse feed from %s", p.url)
}

func (p feedParseError) Cause() error {
	return p.err
}

func (pod *podcast) matchFeed() error {
	if pod.url == "" {
		return errNoFeed
	}

	if pod.ep == nil {
		return errNotScanned
	}

	fp := gofeed.NewParser()
	feed, err := fp.ParseURL(pod.url)
	if err != nil {
		return feedParseError{
			url: pod.url,
			err: err,
		}
	}

	sort.Slice(feed.Items, func(i, j int) bool {
		return feed.Items[i].PublishedParsed.After(*feed.Items[j].PublishedParsed)
	})
	lastDownloaded := pod.mostCurrent()

	match := -1
	for i, item := range feed.Items {
		if len(item.Enclosures) != 1 {
			return errNoEnclosure
		}
		if item.Enclosures[0].URL == "" {
			return fmt.Errorf("no remote URL for episode")
		}

		ok, err := compareRemote(item.Enclosures[0].URL, pod.local, pod.ep[lastDownloaded].filename)
		if err != nil {
			return err
		}
		if ok {
			match = i
			break
		}
	}
	if match == -1 {
		return errNoMatch
	}

	fmt.Fprintf(output, "Successfully matched feed to local directory content; episode #%d is #%d in the feed.\n", lastDownloaded, match)

	for i, n := match-1, lastDownloaded+1; i >= 0; i, n = i-1, n+1 {
		if len(feed.Items[i].Enclosures) != 1 {
			return errNoEnclosure
		}
		if err := pod.downloadEpisode(n, feed.Items[i].Enclosures[0].URL); err != nil {
			return err
		}
	}

	for i, n := 0, pod.mostCurrent(); i < len(feed.Items); i, n = i+1, n-1 {
		if _, ok := pod.ep[n]; !ok {
			if len(feed.Items[i].Enclosures) != 1 {
				return errNoEnclosure
			}
			if err := pod.downloadEpisode(n, feed.Items[i].Enclosures[0].URL); err != nil {
				return err
			}
		}
	}

	return nil
}

func (pod *podcast) downloadEpisode(n int, url string) error {
	fmt.Fprintf(output, "Fetching episode #%d: ", n)

	file, err := downloadFile(pod.local, strconv.Itoa(n), url)
	if err != nil {
		return err
	}
	pod.ep[n] = &episode{filename: file}
	return nil
}

func downloadFile(fs afero.Fs, filenamePrefix, url string) (string, error) {
	fmt.Fprintf(output, "downloading %s...", url)
	client := &http.Client{}
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return "", err
	}
	req.Header.Add("User-Agent", `Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/39.0.2171.27 Safari/537.36`)
	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	filename, err := guessFilename(resp)
	if err != nil {
		return "", err
	}

	filename = filenamePrefix + filepath.Ext(filename)

	out, err := fs.Create(filename)
	if err != nil {
		return "", err
	}
	defer out.Close()

	_, err = io.Copy(out, resp.Body)

	fmt.Fprintf(output, "successfully downloaded %s\n", filename)

	return filename, err
}

func compareRemote(url string, fs afero.Fs, filename string) (bool, error) {
	const chunkSize = 1024 * 64

	fmt.Fprintf(output, "comparing %s to %s...\n", url, filename)
	info, err := fs.Stat(filename)
	if err != nil {
		return false, err
	}

	length := info.Size()

	client := &http.Client{}
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return false, err
	}
	req.Header.Add("User-Agent", `Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/39.0.2171.27 Safari/537.36`)
	resp, err := client.Do(req)
	if err != nil {
		return false, err
	}
	defer resp.Body.Close()

	f, err := fs.Open(filename)
	if err != nil {
		return false, err
	}
	defer f.Close()

	bf := make([]byte, chunkSize)
	br := make([]byte, chunkSize)

	for i := 0; int64(i) < length/chunkSize; i++ {
		if _, err := io.ReadFull(f, bf); err != nil {
			return false, err
		}

		if _, err := io.ReadFull(resp.Body, br); err != nil {
			if err == io.EOF || err == io.ErrUnexpectedEOF {
				return false, nil
			}
			return false, err
		}

		if !bytes.Equal(bf, br) {
			return false, nil
		}
	}

	bf = make([]byte, length%chunkSize)
	br = make([]byte, length%chunkSize)

	if _, err := io.ReadFull(f, bf); err != nil {
		return false, err
	}

	if _, err := io.ReadFull(resp.Body, br); err != nil {
		if err == io.EOF || err == io.ErrUnexpectedEOF {
			return false, nil
		}
		return false, err
	}

	if !bytes.Equal(bf, br) {
		return false, nil
	}

	if n, err := resp.Body.Read(br); n != 0 || err != io.EOF {
		return false, nil
	}

	return true, nil
}

func guessFilename(resp *http.Response) (string, error) {
	filename := resp.Request.URL.Path
	if cd := resp.Header.Get("Content-Disposition"); cd != "" {
		if _, params, err := mime.ParseMediaType(cd); err == nil {
			filename = params["filename"]
		}
	}

	// sanitize
	if filename == "" || strings.HasSuffix(filename, "/") || strings.Contains(filename, "\x00") {
		return "", fmt.Errorf("no filename")
	}

	filename = filepath.Base(filepath.Clean("/" + filename))
	if filename == "" || filename == "." || filename == "/" {
		return "", fmt.Errorf("no filename")
	}

	return filename, nil
}


build.go
// Copyright (C) 2019 Evgeny Kuznetsov (evgeny@kuznetsov.md)
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along tihe this program. If not, see <https://www.gnu.org/licenses/>.

// +build ignore

package main

import (
	"archive/tar"
	"archive/zip"
	"bytes"
	"compress/gzip"
	"flag"
	"fmt"
	"io"
	"log"
	"os"
	"os/exec"
	"regexp"
	"runtime"
	"strconv"
	"time"
)

type packFile struct {
	src string
	dst string
	mod os.FileMode
}

type target struct {
	goos   string
	goarch string
}

const appName = "podsaver"

var (
	filename          string
	binName           string
	goos              string
	goarch            string
	keyID             string = "B5C7C6EAA5B9B41A"
	version           string
	btime             int64
	packFilesRequired = []packFile{
		{src: "LICENSE", dst: "LICENSE", mod: 0644},
		{src: "README.md", dst: "README.md", mod: 0644},
		{src: "SOURCE.txt", dst: "SOURCE.txt", mod: 0644},
	}
	packFiles = []packFile{}
	targets   = []target{
		{goos: "linux", goarch: "386"},
		{goos: "linux", goarch: "amd64"},
		{goos: "linux", goarch: "arm"},
		{goos: "linux", goarch: "arm64"},
		{goos: "windows", goarch: "386"},
		{goos: "windows", goarch: "amd64"},
		{goos: "darwin", goarch: "amd64"},
	}
)

func main() {
	sign := flag.Bool("s", false, "sign binary")
	tar := flag.Bool("t", false, "generate archive")
	buildAll := flag.Bool("a", false, "build archives for all supported systems")
	flag.StringVar(&goos, "goos", runtime.GOOS, "GOOS value")
	flag.StringVar(&goarch, "goarch", runtime.GOARCH, "GOARCH value")
	flag.Parse()
	version = getVersion()
	btime = buildTime()
	if *buildAll {
		*tar = true
	}
	if *tar {
		*sign = true
	}

	if !*buildAll {
		targets = []target{
			{goos: goos, goarch: goarch},
		}
	}
	for _, tgt := range targets {
		packFiles = nil
		packFiles = append(packFiles, packFilesRequired...)
		build(tgt, *sign, *tar)
	}
}

func build(tgt target, sign bool, tar bool) {
	fmt.Printf("Building version %s for %s-%s\n", version, tgt.goos, tgt.goarch)
	fmt.Println("Building as of", time.Unix(btime, 0))

	if tgt.goos == "windows" {
		binName = appName + ".exe"
	} else {
		binName = appName
	}
	os.Setenv("GOOS", tgt.goos)
	os.Setenv("GOARCH", tgt.goarch)
	buildBinary(version, btime)

	signed := false
	if sign {
		signed = signFile(binName, keyID)
	}

	if tar {
		filename = fmt.Sprintf("%s-%s-%s-%s", appName, tgt.goos, tgt.goarch, version)
		if !signed {
			filename = filename + "-unsigned"
		}
		if tgt.goos == "windows" {
			buildZip()
		} else {
			buildTar()
		}
		fmt.Println("archive", filename, "created")
	}
}

func buildBinary(version string, t int64) {
	cmdline := fmt.Sprintf("go build -ldflags \"-X main.version=%s\" -v", version)
	cmd := exec.Command("bash", "-c", cmdline)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	if err := cmd.Run(); err != nil {
		log.Fatalln("failed to build binary")
	}
	setFileTime(binName, t)
	packFiles = append(packFiles, packFile{binName, binName, 0755})
}

func buildZip() {
	filename = filename + ".zip"
	zf, err := os.Create(filename)
	if err != nil {
		log.Fatal(err)
	}
	defer zf.Close()

	zw := zip.NewWriter(zf)
	defer zw.Close()

	for _, f := range packFiles {
		sf, err := os.Open(f.src)
		if err != nil {
			log.Fatal(err)
		}

		info, err := sf.Stat()
		if err != nil {
			log.Fatal(err)
		}

		h, err := zip.FileInfoHeader(info)
		if err != nil {
			log.Fatal(err)
		}

		h.Method = zip.Deflate

		w, err := zw.CreateHeader(h)
		if err != nil {
			log.Fatal(err)
		}

		_, err = io.Copy(w, sf)
		if err != nil {
			log.Fatal(err)
		}

		sf.Close()
	}
}

func buildTar() {
	for i := range packFiles {
		packFiles[i].dst = filename + "/" + packFiles[i].dst
	}
	filename = filename + ".tar.gz"
	fd, err := os.Create(filename)
	if err != nil {
		log.Fatal(err)
	}
	gw, err := gzip.NewWriterLevel(fd, gzip.BestCompression)
	if err != nil {
		log.Fatal(err)
	}
	tw := tar.NewWriter(gw)
	for _, f := range packFiles {
		sf, err := os.Open(f.src)
		if err != nil {
			log.Fatal(err)
		}
		info, err := sf.Stat()
		if err != nil {
			log.Fatal(err)
		}
		h := &tar.Header{
			Name:    f.dst,
			Size:    info.Size(),
			Mode:    int64(f.mod),
			ModTime: info.ModTime(),
		}
		err = tw.WriteHeader(h)
		if err != nil {
			log.Fatal(err)
		}
		_, err = io.Copy(tw, sf)
		if err != nil {
			log.Fatal(err)
		}
		sf.Close()
	}
	err = tw.Close()
	if err != nil {
		log.Fatal(err)
	}
	err = gw.Close()
	if err != nil {
		log.Fatal(err)
	}
	err = fd.Close()
	if err != nil {
		log.Fatal(err)
	}
}

func setFileTime(f string, t int64) {
	cmd := exec.Command("touch", "-t", fmt.Sprint(time.Unix(t, 0).Format("200601021504.05")), f)
	if err := cmd.Run(); err != nil {
		log.Fatalln("failed to set time on", f)
	}
}

func signFile(f string, k string) bool {
	cmd := exec.Command("gpg", "--detach-sign", "--yes", "--passphrase", os.Getenv("GPG_PASSPHRASE"), "--pinentry-mode", "loopback", "-a", "-u", k, f)
	if err := cmd.Run(); err != nil {
		fmt.Println("signing", f, "failed")
		return false
	} else {
		fmt.Println(f, "successfully signed with key", k)
		packFiles = append(packFiles, packFile{binName + ".asc", binName + ".asc", 0644})
		return true
	}
}

func getVersion() string {
	s, err := getString("git", "describe", "--always", "--dirty")
	versionRe := regexp.MustCompile(`^v[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}(\-[0-9]{1,3}\-g[0-9a-f]{5,15})?`)
	if err == nil {
		if versionRe.MatchString(s) {
			return s
		}
	}
	return "unknown"
}

func getString(c string, a ...string) (string, error) {
	cmd := exec.Command(c, a...)
	b, err := cmd.CombinedOutput()
	return string(bytes.TrimSpace(b)), err
}

func buildTime() int64 {
	s, err := getString("git", "show", "-s", "--format=%ct")
	if err == nil {
		if i, e := strconv.ParseInt(s, 10, 64); e == nil {
			return i
		}
	}
	return time.Now().Unix()
}
